Add ElementTag text component attributes

MusicScore opened the pull request:
Implements ElementTag equivalents for `<&hover[<text>].type[<type>]>ELEMENT<&end_hover>`, `<&click[<text>].type[<type>]>ELEMENT<&end_click>`, and `<&insertion[<text>]>ELEMENT<&end_insertion>`.

Some extra utility stuff had to be thrown in so that certain convenience features (`<ElementTag.item_hover[<item>/<text>]>`) could work, and some extra arguably unnecessary processing had to be done for automatic output to look reasonable to the average user (`<ElementTag.entity_hover[<entity>/<text>]>`).

I did not fully test all of the changes except for basic functionality. The checklist notes which I have tested regarding error handling or issues:

- [x] `<ElementTag.text_hover[<text>]>`
- [x] `<ElementTag.item_hover[<item>/<text>]>`
- [x] `<ElementTag.entity_hover[<entty>/<text>]>`
- [x] `<ElementTag.run_on_click[<command>]>`
- [x] `<ElementTag.suggest_on_click[<command>]>`
- [x] `<ElementTag.page_on_click[<#>]>`
- [x] `<ElementTag.url_on_click[<url>]>`
- [x] `<ElementTag.file_on_click[<filepath>]>`
- [x] `<ElementTag.insert_on_shift_click[<text>]>`

Fixes #35

MusicScore committed:
Add various JSON text attributes to ElementTag

Implement <&hover[]>, <&click[]>, and <&insertion[]> as ElementTag properties

Add smart item parsing for <ElementTag.item_hover[<item>/<text>]>

Add smart entity parsing for <ElementTag.entity_hover[<entity>/<text>]>

Add <ElementTag.to_json> and <ElementTag.from_json> for easier conversions

Add an additional helper method in ItemHelper to get a valid JSON string for <ElementTag.item_hover[]>

Add <ItemTag.unmodified_json> to complement the new ItemHelper#getUnmodifiedJsonString(ItemStack) method

MusicScore committed:
Remove accidental debug

MusicScore committed:
Additional minor fixes

Don't let every entity's type default to "Unknown"

Fix small typo in debug for <ElementTag.insert_on_shift_click[]>

MusicScore committed:
Remove to/from JSON attributes

These are implemented upstream, under different names

MusicScore committed:
Minor optimization to <ElementTag.entity_hover[<entity>/<text>]>

If the entity type doesn't need to be titlecased, don't perform the titlecasing

ghost commented:
I'm aware that the upstream has implemented this PR in a different fashion. However, I still strongly believe that this PR is still useful; the user should not have to be forced to input an ItemTag for the `SHOW_ITEM` hover action, or be forced to input an EntityTag for `SHOW_ENTITY` if they wish to input their own custom JSON text (for any reason whatsoever). Regardless, the upstream seems to not fully understand what inputs are actually required for either; I am not going to pull back this PR for whatever the upstream has done.

EDIT: Also, `.suggest_on_run[<command>]` is just less typing than `on_click[<command>].type[SUGGEST_COMMAND]`. Generally speaking, most of the tags proposed by this PR are significantly shorter than the upstream's implementation, and (in my opinion) are implemented better.

MusicScore committed:
Match upstream methodology

Use the methods that upstream uses for convenience and general future compatibility

Make item/entity input for "item_hover", "entity_hover", "on_hover[<text>].type[SHOW_ITEM]", and "on_hover[<text>].type[SHOW_ENTITY]" optional

Removed <ItemTag.unmodified_json> as I cannot get its original intention to work properly

Removed "<ElementTag.insert_on_shift_click>" since "<ElementTag.with_insertion>" was added upstream

ghost commented:
Tested to work on Paper 1.14.4
